### PR TITLE
#34: Make Laplacian a  Kokkos Kernels CRS matrix and use Gauss Seidel linear solve

### LINKED ISSUE #34: Make Laplacian a  Kokkos Kernels BlockCRS matrix and use Gauss Seidel linear solve
This is in order to be less dependent on a single part of the application currently implemented by means of a naïve conjugate gradient solve.

### PR DESCRIPTION
**Pull request by ppebay**
Fixes #34

### CODE REVIEW

**Comment by ppebay** [comment]
@lifflander @pierrepebay this is ready for review

File: cppfd/allTests.cpp [hunk 1]
```diff
@@ -26,50 +26,69 @@ struct solver_test : testing::Test{
 
     // define boundary conditions
     std::map<std::string, double> velocity_values = {
-                                                    {"u_top", 1.0},
```
Comment (lifflander) [inline cppfd/allTests.cpp:29]: Was this spacing change intentional?

File: cppfd/solver.cpp [hunk 2]
```diff
@@ -1,29 +1,32 @@
...
 
+#include <Kokkos_Core.hpp>
+
 #include <KokkosBlas1_axpby.hpp>
+#include <KokkosBlas1_nrm2_squared.hpp>
 #include <KokkosBlas1_dot.hpp>
-#include <KokkosBlas2_gemv.hpp>
-#include <KokkosBlas3_gemm.hpp>
+
+#include "KokkosSparse_CrsMatrix.hpp"
```
Comment (lifflander) [inline cppfd/solver.cpp:9]: Can you make all the Kokkos includes use `<`,`>` for consistency?

File: cppfd/solver.cpp [hunk 3]
```diff
@@ -129,91 +151,158 @@ void Solver::solve(stopping_point s_p, linear_solver l_s,adaptative_time_step at
...
+  Kokkos::View<double*> values("values", nnz);
+
+  // iterate m*n over cells to construct mn*mn Laplacian
+  uint64_t idx = 0;
+  for(uint64_t j = 0; j < n; j++){
+    for(uint64_t i = 0; i < m; i++){
+      uint64_t k = this->mesh.cartesian_to_index(i, j, m, n);
+      bool first_in_row = true;
+      // assign below diagonal entries when relevant
+      if(j > 0)
```
Comment (lifflander) [inline cppfd/solver.cpp:234]: The indentation and formatting here looks like it's off.

File: cppfd/solver.cpp [hunk 4]
```diff
@@ -129,91 +151,158 @@ void Solver::solve(stopping_point s_p, linear_solver l_s,adaptative_time_step at
...
+  const uint64_t n = this->mesh.get_n_cells_y();
+  const uint64_t mn = m * n;
+  const uint64_t nnz = 5 * mn - 2 * (m + n);
+  Kokkos::View<uint64_t*> row_ptrs("row pointers", mn + 1);
+  Kokkos::View<uint64_t*> col_ids("column indices", nnz);
+  Kokkos::View<double*> values("values", nnz);
+
+  // iterate m*n over cells to construct mn*mn Laplacian
+  uint64_t idx = 0;
+  for(uint64_t j = 0; j < n; j++){
```
Comment (lifflander) [inline cppfd/solver.cpp:229]: This might be easier to read as a separate function that produces the sparse matrix.

**Review by lifflander — APPROVED** [review]
Overall this looks great!. I have a few minor suggestions for improvement. Some of the spacing and indentation seems to be off. Your editor might be inserting tabs or spaces that is changing the alignment.

File: cppfd/solver.cpp (hunk 4 above)
Comment (ppebay) [reply cppfd/solver.cpp:252]: @lifflander I made a refactoring following your comment, but also extended it in order to expand the use of the Kokkos arithmetic traits `zero` and `one`, can you please have a look? Thanks

**Review by pierrepebay — APPROVED** [review]
Numerical results appear to be correct, compared to openFoam outputs.
